File: python/src/report_processing/report_parser.py
"""Filename parser for the Python implementation."""

from report_processing.report_domain import ReportCategory, ReportMetadata, ReportType


class ReportParser:
    """Validates and parses filenames shaped as ``CAT_TIPO_nombre.txt``."""

    def parse(self, file_name: str) -> ReportMetadata:
        if not file_name or not file_name.strip():
            raise ValueError("The report file name cannot be null or blank.")
        if not file_name.endswith(".txt"):
            raise ValueError(f"The report file name must end with .txt: {file_name}")

        tokens = file_name[:-4].split("_")
        if len(tokens) < 3:
            raise ValueError(
                f"The report file name must follow CAT_TIPO_nombre.txt: {file_name}"
            )

        return ReportMetadata(
            category=ReportCategory.from_code(tokens[0]),
            report_type=ReportType.from_code(tokens[1]),
            file_name=file_name,
        )

File: java/src/main/java/com/unir/reuse/reports/ReportParser.java
package com.unir.reuse.reports;

/**
 * Parses filenames that follow the contract described in the activity:
 * {@code CAT_TIPO_nombre.txt}. The parser is intentionally strict because a
 * malformed filename prevents the factory selection from being deterministic.
 */
public final class ReportParser {

    public ReportMetadata parse(final String fileName) {
        if (fileName == null || fileName.isBlank()) {
            throw new IllegalArgumentException("The report file name cannot be null or blank.");
        }
        if (!fileName.endsWith(".txt")) {
            throw new IllegalArgumentException("The report file name must end with .txt: " + fileName);
        }

        String fileStem = fileName.substring(0, fileName.length() - 4);
        String[] tokens = fileStem.split("_");

        if (tokens.length < 3) {
            throw new IllegalArgumentException(
                "The report file name must follow CAT_TIPO_nombre.txt: " + fileName
            );
        }

        ReportCategory category = ReportCategory.fromCode(tokens[0]);
        ReportType type = ReportType.fromCode(tokens[1]);

        return new ReportMetadata(category, type, fileName);
    }
}

